### ISSUE TITLE
webpack --entry 'webpack' --output-filename './webpack.bundle.js'

### ISSUE THREAD

**Issue by abhishekisnot** [webpack/webpack#1434]
```javascript
webpack --entry 'webpack' --output-filename './webpack.bundle.js'
```
webpack should be able to bundle itself.

Motivations: 
Using [webpack with backend apps](http://jlongster.com/Backend-Apps-with-Webpack--Part-I)
errors on [dynamic requires in node-dir](https://github.com/fshost/node-dir/issues/22)

**Comment by ivan-kleshnin** [comment]
@abhishekisnot as an opportunity to ask... Why you need Webpack for backend apps?
James Longster forgot to mention in that article actual benefits one gonna get from this... 
Except hot module replacement which I personally never needed. 

Bundles? Backend has a notion of modules. Async loads? Sync works arguably better for server side. 
I see only tons of drawbacks for now and I'd really like to see benefits as well.

**Comment by rgranger** [comment]
:+1: 

Same issue

I need to bundle my NodeJS server with webpack, and my server is using webpack-dev-server and webpack-hmr when NODE_ENV=development

So I have to had webpack to externals (and every other modules with the same issue), which is quite annoying. : 

    webpackConfig.externals = {
      'yargs': 'commonjs yargs',
      'socket.io': 'commonjs socket.io',
      'webpack': 'commonjs webpack',
      'html-webpack-plugin': 'commonjs html-webpack-plugin',
      'any-promise': 'commonjs any-promise',
      'natives': 'commonjs natives',
       'chokidar': 'commonjs chokidar'
    }

**Comment by jhnns** [comment]
When webpack is used to bundle node.js apps (usually necessary for SSR), you should always exclude `node_modules` by configuring the `externals`. This speeds up the build significantly.

Personally, I don't see a use-case where you'd want to bundle webpack itself.

**Comment by rgranger** [comment]
Yes, you are right, for most use cases.

My specific use case is the following :

Our NodeJS server is embedded into machines that doesn't have necessarily internet access. Thus, when we upload a new software version for our machines, our customers have to download the new version, and execute an updater.

Now, we also provide a "simulator", which simply replicates what the machine do (minus the hardware part / video processing). The simulator is an Electron software that simply launch our NodeJS server with a --simulation option.
The simulator has all the NodeJS server releases so the user can choose the version of the machines he wants to simulate.

For this we need the server releases to be bundled and tagged with their version.

In order to work, the bundle contains the following :
* The bundled client
* The bundled server
* The node_modules that can't be bundled

In order to have an efficient package, I have to remove from the node_modules all the modules that are already bundled in the client, and all the devDependencies, to keep only the "useful node_modules"

If I don't do this step, and just put in all the node_modules, I have a package > 1Go. When I do, I have a package of ~5Mo.

It would be far simpler if I hadn't had to worry about including the node_modules folder with each releases, and my bundled server file worked on its own (like the client bundle).

Then my server bundle would be bigger, but i wouldn't have to worry about the modules that "can't be compiled"

**Comment by jhnns** [comment]
Thanks for providing a use-case.

I also stumbled upon a use-case: Initial start-up performance. Since webpack is a CLI tool, start-up performance is really important. Node's sync require(), however, can really slow things down. [This graph shows](https://cloud.githubusercontent.com/assets/781746/24209045/7679a5fc-0f25-11e7-9d59-1fb739a64cf2.jpg) that during the first second, we're just spending requiring things.

We could (probably) speed up things by bundling webpack (and loaders). That would probably decrease the time spend in require()

**Comment by osdevisnot** [comment]
Perfect use cases. In addition to that if such bundling is possible, it would drastically reduce installation time for a typical CLI apps. 

PS: my apologies @ivan-kleshnin. Don't know how I missed your  comment before.

**Comment by osdevisnot** [comment]
Also see this https://github.com/angular/angular-cli/issues/3199

**Comment by ivan-kleshnin** [comment]
@osdevisnot no problem. Thanks!

**Comment by peterebden** [comment]
I've been having a look into this and it nearly works, at least for my use case. The main remaining issue seems to be loaders; if we could supply loaders as objects (ala how Babel presets can be supplied either way) then I think Webpack would work as a bundle, at least for simple cases.

I've tried to make the change myself but am a bit lost about where's best for it; I'm not familiar with Webpack's internals - or possibly it should be in loader-runner?

**Comment by jhnns** [comment]
Yeah, loaders are still a problem because they are loaded with a dynamic `require()` expression. This can likely be solved with the [`ContextReplacementPlugin`](https://webpack.js.org/plugins/context-replacement-plugin/), but I haven't looked into it yet.

To the issue in general: I was able to bundle webpack with webpack in order to build a benchmark for browser vendors. You can take a look here: https://github.com/jhnns/webpack-browser-benchmark. That's without loaders, though.
